File: src/main/java/mediathek/gui/abo/AboInformationController.java
package mediathek.gui.abo;

import javafx.fxml.FXML;
import javafx.scene.control.Label;
import mediathek.config.Daten;
import mediathek.gui.messages.TimerEvent;
import mediathek.javafx.tool.JavaFxUtils;
import net.engio.mbassy.listener.Handler;

public class AboInformationController {
    @FXML
    private Label totalAbos;

    @FXML
    private Label activeAbos;

    @FXML
    private Label inactiveAbos;

    private int oldSize = -1;
    private long oldActive = -1;
    private long oldInactive = -1;
    private Daten daten;

    public void startListener() {
        this.daten = Daten.getInstance();
        daten.getMessageBus().subscribe(this);
    }

    private void updateTotalDisplay(final int gesamt) {
        if (gesamt == 1) {
            totalAbos.setText("Gesamt: 1 Abo");
        } else {
            totalAbos.setText(String.format("Gesamt: %d Abos", gesamt));
        }
    }

    private void updateDisplayText() {
        final var listeAbo = daten.getListeAbo();
        final long _activeAbos = listeAbo.activeAbos();
        final long _inactiveAbos = listeAbo.inactiveAbos();
        final int gesamt = listeAbo.size();

        if (gesamt != oldSize) {
            updateTotalDisplay(gesamt);

            oldSize = gesamt;
        }

        if (_activeAbos != oldActive) {
            activeAbos.setText(String.format("%d eingeschaltet", _activeAbos));
            oldActive = _activeAbos;
        }

        if (_inactiveAbos != oldInactive) {
            inactiveAbos.setText(String.format("%d ausgeschaltet", _inactiveAbos));
            oldInactive = _inactiveAbos;
        }
    }

    @Handler
    private void handleTimerEvent(TimerEvent e) {
        JavaFxUtils.invokeInFxThreadAndWait(this::updateDisplayText);
    }
}

File: src/main/java/mediathek/gui/abo/AboInformationController.kt
package mediathek.gui.abo

import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.control.Label
import mediathek.config.Daten
import mediathek.daten.abo.DatenAbo
import mediathek.gui.messages.AboListChangedEvent
import mediathek.javafx.tool.JavaFxUtils
import mediathek.tool.MessageBus
import net.engio.mbassy.listener.Handler
import java.net.URL
import java.util.*

class AboInformationController : Initializable {
    @FXML
    private lateinit var totalAbos: Label

    @FXML
    private lateinit var activeAbos: Label

    @FXML
    private lateinit var inactiveAbos: Label

    private fun updateDisplayText() {
        val listeAbo = Daten.getInstance().listeAbo
        val numAbos = listeAbo.size

        if (numAbos == 1) {
            totalAbos.text = "Gesamt: 1 Abo"
        } else {
            totalAbos.text = "Gesamt: $numAbos Abos"
        }

        activeAbos.text = "${numActiveAbos()} eingeschaltet"
        inactiveAbos.text = "${numInactiveAbos()} ausgeschaltet"
    }

    /**
     * Get the number of abos which are active and used.
     *
     * @return num of used abos
     */
    private fun numActiveAbos(): Long {
        return Daten.getInstance().listeAbo.stream().filter { obj: DatenAbo -> obj.isActive }.count()
    }

    /**
     * Get the number of abos which are created but offline.
     *
     * @return number of abos which are offline
     */
    private fun numInactiveAbos(): Long {
        return Daten.getInstance().listeAbo.stream().filter { abo: DatenAbo -> !abo.isActive }.count()
    }

    @Handler
    @Suppress("UNUSED_PARAMETER")
    private fun handleAboChangedEvent(e: AboListChangedEvent) {
        JavaFxUtils.invokeInFxThreadAndWait { updateDisplayText() }
    }

    override fun initialize(location: URL?, resources: ResourceBundle?) {
        MessageBus.messageBus.subscribe(this)
        updateDisplayText()
    }
}